06_gpu_and_ml/comfyui/comfy_ui.py

```diff
@@ -1,140 +1,157 @@
-        "cd /root/ComfyUI-to-Python-Extension && pip install -r requirements.txt",
     )
-    .pip_install(
-        "httpx",
-        "requests",
-        "tqdm",
+    .pip_install("httpx", "tqdm", "websocket-client")
+    .copy_local_file(
+        pathlib.Path(__file__).parent / "model.json", "/root/model.json"
+    )
+    .copy_local_file(
+        pathlib.Path(__file__).parent / "helpers.py", "/root/helpers.py"
     )
-    .run_function(download_checkpoints)
-    .run_function(download_plugins)
 )
+
 app = modal.App(
-    name="example-comfy-ui", image=image
-)  # Note: prior to April 2024, "app" was called "stub"
+    name="example-comfyui",
+)
 
-# ## Start the ComfyUI server
-#
-# Inside the container, we will run the ComfyUI server and execution queue on port 8188. Then, we
-# wrap this function in the `@web_server` decorator to expose the server as a web endpoint.
-#
-# For ASGI-compatible frameworks, you can also use Modal's `@asgi_app` decorator.
+# You can define custom checkpoints, plugins, and more in the `model.json` file in this directory.
 
 
-@app.function(
-    gpu="any",
-    # Allows 100 concurrent requests per container.
+# ComfyUI-specific code lives in `helpers.py`.
+# This includes functions like downloading checkpoints/plugins to the right directory on the ComfyUI server.
+with comfyui_image.imports():
+    from helpers import (
+        connect_to_local_server,
+        download_to_comfyui,
+        get_images,
+    )
+
+
+# Here we use Modal's class syntax to build the image (with our custom checkpoints/plugins).
+@app.cls(
     allow_concurrent_inputs=100,
-    # Restrict to 1 container because we want our ComfyUI session state
-    # to be on a single container.
-    concurrency_limit=1,
-    keep_warm=1,
+    gpu="any",
+    image=comfyui_image,
     timeout=1800,
+    container_idle_timeout=300,
+    mounts=[
+        modal.Mount.from_local_file(
+            pathlib.Path(__file__).parent / "workflow_api.json",
+            "/root/workflow_api.json",
+        )
+    ],
 )
-@modal.web_server(8188, startup_timeout=30)
-def web():
-    cmd = "python main.py --dont-print-server --multi-user --listen --port 8188"
-    subprocess.Popen(cmd, shell=True)
+class ComfyUI:
+    @modal.build()
+    def download_models(self):
+        models = json.loads(
+            (pathlib.Path(__file__).parent / "model.json").read_text()
+        )
+        for m in models:
+            download_to_comfyui(m["url"], m["path"])
+
+    def _run_comfyui_server(self, port=8188):
+        cmd = f"python main.py --dont-print-server --listen --port {port}"
+        subprocess.Popen(cmd, shell=True)
+
+    @modal.web_server(8188, startup_timeout=30)
+    def ui(self):
+        self._run_comfyui_server()
+
+    # When you run `modal serve 06_gpu_and_ml/comfyui/comfy_ui.py`, you'll see a `ComfyUI.ui` link to interactively develop your ComfyUI workflow that has the custom checkpoints/plugins loaded in.
```

Comment:
note that the linter added spaces in front of the comments, so they aren't rendered properly

Reply:
what should we do about the ruff ci check then?

Reply:
unfortunately, we just have to reorganize the text.

i have looked into ways around this, e.g. adding methods to classes after they are defined, but haven't been able to find a way to do it